# Module 3: Exercise B

Always load the packages first:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Read Data to Pandas DataFrame

Let's start by reading the full data set to a pandas DataFrame. Note the last column __y__ has the campaign outcome (i.e., "yes" or "no" to the term deposit).

>__Task 1__
>
>Import and explore the data file "bank_campaign.csv" (Make sure the file is in the same folder with this Jupyter notebook)

In [2]:
bank_df = ...

Take a look at the first 10 rows to make an intial check of the data import:

In [ ]:
...

Also, `df.shape` tells you how many rows and columns this data set has:

In [4]:
bank_df.shape

(4521, 17)

The data has 4521 rows and 17 columns corresponding to 4521 observations and 16 variables/features. (__y__ column  is the campaign outcome.)

A more comprehensive check is the `info()` method. Not only does it show the columns and rows, it also shows the data types and memory usage.

In [5]:
bank_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4521 entries, 0 to 4520
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   age        4521 non-null   float64
 1   job        4483 non-null   object 
 2   marital    4521 non-null   object 
 3   education  4521 non-null   object 
 4   default    4521 non-null   object 
 5   balance    4424 non-null   float64
 6   housing    4521 non-null   object 
 7   loan       4521 non-null   object 
 8   contact    3197 non-null   object 
 9   day        4521 non-null   float64
 10  month      4521 non-null   object 
 11  duration   4521 non-null   float64
 12  campaign   4521 non-null   float64
 13  pdays      4463 non-null   float64
 14  previous   4521 non-null   float64
 15  poutcome   4521 non-null   object 
 16  y          4521 non-null   object 
dtypes: float64(7), object(10)
memory usage: 600.6+ KB


---

## Select Columns or Rows

Let's assume we want to select the __age__ column. We run:

In [6]:
bank_df['age']

0       30.0
1       33.0
2       35.0
3       30.0
4       59.0
        ... 
4516    33.0
4517    57.0
4518    57.0
4519    28.0
4520    44.0
Name: age, Length: 4521, dtype: float64

The result is NOT a pd.DataFrame but a pd.Series object which is one column of a pd.DataFrame. In other words, __a pd.DataFrame consists of many pd.Series objects.__

Note that in Python, __indices start from 0 (not 1)__. Therefore, if we want the 5th observation, we can write:

In [7]:
bank_df['age'][4]

59.0

Alternatively, we can use index based location (`df.iloc`) to select rows or columns. This is called "slicing".

__age__ is the first column, so the index is 0:

In [8]:
bank_df.iloc[:,0]

0       30.0
1       33.0
2       35.0
3       30.0
4       59.0
        ... 
4516    33.0
4517    57.0
4518    57.0
4519    28.0
4520    44.0
Name: age, Length: 4521, dtype: float64

Similarly, to get the 5th observation, we write:

In [9]:
bank_df.iloc[4,0]

59.0

>__Task 2__
>
>What is the job of the client in the 102nd row? Use either `df[colname][rowname]` or `df.iloc` methods to find it.

In [ ]:
...

### Select More Than One Column

If we want to select more than one column, we must specify the names as another list. Let's assume we want to select the __age__, __job__, __marital__, __education__, and __housing__ columns, we can:

1. Plug in the name of the columns as:

In [11]:
cnames = ['age','job', 'marital', 'education','housing']
cnames

['age', 'job', 'marital', 'education', 'housing']

In [12]:
bank_df[cnames]

age            job  marital  education housing
0     30.0     unemployed  married    primary      no
1     33.0       services  married  secondary     yes
2     35.0     management   single   tertiary     yes
3     30.0     management  married   tertiary     yes
4     59.0    blue-collar  married  secondary     yes
...    ...            ...      ...        ...     ...
4516  33.0       services  married  secondary     yes
4517  57.0  self-employed  married   tertiary     yes
4518  57.0     technician  married  secondary      no
4519  28.0    blue-collar  married  secondary      no
4520  44.0   entrepreneur   single   tertiary     yes

[4521 rows x 5 columns]

2. Or directly define the names without creating a list:

In [13]:
bank_df[['age','job', 'marital', 'education','housing']]

age            job  marital  education housing
0     30.0     unemployed  married    primary      no
1     33.0       services  married  secondary     yes
2     35.0     management   single   tertiary     yes
3     30.0     management  married   tertiary     yes
4     59.0    blue-collar  married  secondary     yes
...    ...            ...      ...        ...     ...
4516  33.0       services  married  secondary     yes
4517  57.0  self-employed  married   tertiary     yes
4518  57.0     technician  married  secondary      no
4519  28.0    blue-collar  married  secondary      no
4520  44.0   entrepreneur   single   tertiary     yes

[4521 rows x 5 columns]

Alternatively, we can use `iloc`:

In [14]:
bank_df.iloc[:,[0,1,2,3,6]]

age            job  marital  education housing
0     30.0     unemployed  married    primary      no
1     33.0       services  married  secondary     yes
2     35.0     management   single   tertiary     yes
3     30.0     management  married   tertiary     yes
4     59.0    blue-collar  married  secondary     yes
...    ...            ...      ...        ...     ...
4516  33.0       services  married  secondary     yes
4517  57.0  self-employed  married   tertiary     yes
4518  57.0     technician  married  secondary      no
4519  28.0    blue-collar  married  secondary      no
4520  44.0   entrepreneur   single   tertiary     yes

[4521 rows x 5 columns]

>__Task 3__
>
>Create a dataset of __marital__, __education__, __housing__, and __y__, using both `df[colname][rowname]` and `df.iloc` methods

In [ ]:
...

In [ ]:
...

---

## Find Missing Values

In our data file, missing values are coded as empty fields. When imported to the DataFrame, empty fields are automatically coded as missing value.

The `isna()` method determines for each cell whether it is a missing value or not (`True` for missing value).

In [17]:
bank_df.isna()

age    job  marital  education  default  balance  housing   loan  \
0     False  False    False      False    False    False    False  False   
1     False  False    False      False    False    False    False  False   
2     False  False    False      False    False    False    False  False   
3     False  False    False      False    False     True    False  False   
4     False  False    False      False    False    False    False  False   
...     ...    ...      ...        ...      ...      ...      ...    ...   
4516  False  False    False      False    False    False    False  False   
4517  False  False    False      False    False    False    False  False   
4518  False  False    False      False    False    False    False  False   
4519  False  False    False      False    False    False    False  False   
4520  False  False    False      False    False    False    False  False   

      contact    day  month  duration  campaign  pdays  previous  poutcome  \
0       False  False  False     False     False  False     False     False   
1       False  False  False     False     False  False     False     False   
2       False  False  False     False     False  False     False     False   
3        True  False  False     False     False  False     False     False   
4        True  False  False     False     False  False     False     False   
...       ...    ...    ...       ...       ...    ...       ...       ...   
4516    False  False  False     False     False  False     False     False   
4517     True  False  False     False     False  False     False     False   
4518    False  False  False     False     False  False     False     False   
4519    False  False  False     False     False  False     False     False   
4520    False  False  False     False     False  False     False     False   

          y  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
4516  False  
4517  False  
4518  False  
4519  False  
4520  False  

[4521 rows x 17 columns]

However, we would like to see for each column whether there are any missing values. Try again with the `sum()` method to sum the Boolean values up by columns.

In [18]:
bank_df.isna().sum()

age             0
job            38
marital         0
education       0
default         0
balance        97
housing         0
loan            0
contact      1324
day             0
month           0
duration        0
campaign        0
pdays          58
previous        0
poutcome        0
y               0
dtype: int64

The `sum()` method can also sum up along the columns (`axis=1`), which we can use to check the number of missing values in any row.

>__Task 4__
>
>Find the number of missing values for each row

In [ ]:
...

Some missing values are subtle: it may have value of 0 or string NA. This data set has "unknown" in the __education__ and __poutcome__ columns.

In [20]:
bank_df[bank_df['education']=='unknown']

age          job   marital education default  balance housing loan  \
20    38.0   management  divorced   unknown      no      0.0     yes   no   
27    67.0      retired   married   unknown      no    696.0      no   no   
49    61.0       admin.   married   unknown      no   4629.0     yes   no   
132   43.0  blue-collar   married   unknown     yes   -715.0     yes   no   
133   48.0       admin.   married   unknown      no      0.0     yes   no   
...    ...          ...       ...       ...     ...      ...     ...  ...   
4337  73.0      retired   married   unknown      no      NaN      no   no   
4342  50.0  blue-collar   married   unknown      no   2284.0     yes   no   
4428  48.0  blue-collar   married   unknown      no    439.0     yes   no   
4471  59.0   management   married   unknown      no   3534.0      no   no   
4508  42.0       admin.   married   unknown      no    642.0     yes  yes   

        contact   day month  duration  campaign  pdays  previous poutcome    y  
20     cellular  18.0   nov      96.0       2.0   -1.0       0.0  unknown   no  
27    telephone  17.0   aug     119.0       1.0  105.0       2.0  failure   no  
49     cellular  27.0   jan     181.0       1.0   92.0       1.0  success  yes  
132         NaN  30.0   may      67.0       3.0   -1.0       0.0  unknown   no  
133    cellular   8.0   may      85.0       1.0  168.0       2.0  failure   no  
...         ...   ...   ...       ...       ...    ...       ...      ...  ...  
4337  telephone  16.0   oct     434.0       1.0   57.0       1.0  failure  yes  
4342  telephone  31.0   jul    1088.0      17.0   -1.0       0.0  unknown  yes  
4428   cellular   2.0   feb     212.0       1.0  249.0       2.0    other   no  
4471   cellular  21.0   nov     216.0       4.0   -1.0       0.0  unknown   no  
4508        NaN  16.0   may     509.0       2.0   -1.0       0.0  unknown   no  

[187 rows x 17 columns]

We can replace "unknown" with "np.NaN":

In [21]:
bank_df['education'] = bank_df['education'].replace('unknown',np.NaN)

Check __education__ again to see if it still has value "unknown".

In [22]:
bank_df[bank_df['education']=='unknown']

Empty DataFrame
Columns: [age, job, marital, education, default, balance, housing, loan, contact, day, month, duration, campaign, pdays, previous, poutcome, y]
Index: []

>__Task 5__
>
>- Replace "unknown" with "np.NaN" for the __poutcome__ column
>- Can you make sure the __poutcome__ column does not have "unknown" anymore?

In [ ]:
...

In [ ]:
...

---

## Remove Missing Values

To remove missing values, we can use `pd.DataFrame.dropna()`. See [pd.DataFrame.dropna()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.dropna.html) for the important parameters.

Check again the number of missing values to decide how to deal with each column/variable:

In [25]:
bank_df.isna().sum()

age             0
job            38
marital         0
education     187
default         0
balance        97
housing         0
loan            0
contact      1324
day             0
month           0
duration        0
campaign        0
pdays          58
previous        0
poutcome     3705
y               0
dtype: int64

__job__ and __education__ are categorical features that have a relatively small portion of missing values. So, a convenient solution is to remove the rows with missing values.

In [26]:
bank_df.dropna(subset=['job','education'], inplace=True)

>__Task 6__
>
>- Remove the rows where all of the fields are NA
>- Show the count of the missing values for each column

In [ ]:
...

Two columns have a large portion of missing values. Can you find them? Dropping rows for these columns will cause significant loss of observations. Therefore, the better option is to drop the columns.

>__Task 7__
>
>- Remove the columns with non-missing value < 4000, using the `df.dropna()` method (Tip: do not use `inplace=True` if you want to experiment first)
>- Count the NAs for each column. How many NAs exist in your new data set?

In [ ]:
...

---

## Replace Missing Values with Proper Values

The `Dataframe.fillna()` function fills the missing values with a given method. See [pandas.DataFrame.fillna](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.fillna.html) for the important parameters.

Let's start with replacing values with their averages. Since each column has a different average, we need to do it separately. For example, if we want to replace NAs in __pdays__ with its average, we need to calculate the average first:

In [29]:
bank_df['pdays'].mean()

39.47156954887218

Now, let's replace the NAs with the above average:

In [30]:
# temp for temporary data
temp = bank_df
temp['pdays'] = temp['pdays'].fillna(temp['pdays'].mean())
temp

age            job  marital  education default  balance housing loan  \
0     30.0     unemployed  married    primary      no   1787.0      no   no   
1     33.0       services  married  secondary      no   4789.0     yes  yes   
2     35.0     management   single   tertiary      no   1350.0     yes   no   
3     30.0     management  married   tertiary      no      NaN     yes  yes   
4     59.0    blue-collar  married  secondary      no      0.0     yes   no   
...    ...            ...      ...        ...     ...      ...     ...  ...   
4516  33.0       services  married  secondary      no   -333.0     yes   no   
4517  57.0  self-employed  married   tertiary     yes  -3313.0     yes  yes   
4518  57.0     technician  married  secondary      no    295.0      no   no   
4519  28.0    blue-collar  married  secondary      no   1137.0      no   no   
4520  44.0   entrepreneur   single   tertiary      no   1136.0     yes  yes   

       day month  duration  campaign  pdays  previous   y  
0     19.0   oct      79.0       1.0   -1.0       0.0  no  
1     11.0   may     220.0       1.0  339.0       4.0  no  
2     16.0   apr     185.0       1.0  330.0       1.0  no  
3      3.0   jun     199.0       4.0   -1.0       0.0  no  
4      5.0   may     226.0       1.0   -1.0       0.0  no  
...    ...   ...       ...       ...    ...       ...  ..  
4516  30.0   jul     329.0       5.0   -1.0       0.0  no  
4517   9.0   may     153.0       1.0   -1.0       0.0  no  
4518  19.0   aug     151.0      11.0   -1.0       0.0  no  
4519   6.0   feb     129.0       4.0  211.0       3.0  no  
4520   3.0   apr     345.0       2.0  249.0       7.0  no  

[4311 rows x 15 columns]

>__Task 8__
>
>Fill the missing values with the feature means for __balance__ column
>
>Assign the data set to "temp" and work on "temp"

In [ ]:
...
temp

Only for learning purposes, we can replace the missing values in a row with the next available data by using `method='bfill'` as below. Note here `inplace=False` tells `fillna` not to make any actual change to the DataFrame. You can also simply remove this argument since `False` is default.

In [32]:
bank_df.fillna(method='bfill', limit=10, inplace=False)

age            job  marital  education default      balance housing  \
0     30.0     unemployed  married    primary      no  1787.000000      no   
1     33.0       services  married  secondary      no  4789.000000     yes   
2     35.0     management   single   tertiary      no  1350.000000     yes   
3     30.0     management  married   tertiary      no  1413.494906     yes   
4     59.0    blue-collar  married  secondary      no     0.000000     yes   
...    ...            ...      ...        ...     ...          ...     ...   
4516  33.0       services  married  secondary      no  -333.000000     yes   
4517  57.0  self-employed  married   tertiary     yes -3313.000000     yes   
4518  57.0     technician  married  secondary      no   295.000000      no   
4519  28.0    blue-collar  married  secondary      no  1137.000000      no   
4520  44.0   entrepreneur   single   tertiary      no  1136.000000     yes   

     loan   day month  duration  campaign  pdays  previous   y  
0      no  19.0   oct      79.0       1.0   -1.0       0.0  no  
1     yes  11.0   may     220.0       1.0  339.0       4.0  no  
2      no  16.0   apr     185.0       1.0  330.0       1.0  no  
3     yes   3.0   jun     199.0       4.0   -1.0       0.0  no  
4      no   5.0   may     226.0       1.0   -1.0       0.0  no  
...   ...   ...   ...       ...       ...    ...       ...  ..  
4516   no  30.0   jul     329.0       5.0   -1.0       0.0  no  
4517  yes   9.0   may     153.0       1.0   -1.0       0.0  no  
4518   no  19.0   aug     151.0      11.0   -1.0       0.0  no  
4519   no   6.0   feb     129.0       4.0  211.0       3.0  no  
4520  yes   3.0   apr     345.0       2.0  249.0       7.0  no  

[4311 rows x 15 columns]

---

## Create Duplicated Data

The original data does not have any row duplicating altogether. So, we will first create an artificial data set by repeating the same data twice:

In [33]:
bank_duplicated = pd.concat([bank_df]*2, ignore_index=True)

print(bank_df.shape, bank_duplicated.shape)

(4311, 15) (8622, 15)


---

## Find Duplicates

`DataFrame.duplicated()` returns a Boolean Series (`True` for duplicate row). 

In [34]:
bank_duplicated.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
8617     True
8618     True
8619     True
8620     True
8621     True
Length: 8622, dtype: bool

>__Task 9__
>
>Check if __duration__ column is duplicated and count the number of duplicates

In [ ]:
...

In [ ]:
...

---

## Remove Duplicate Rows

The above method just locates the duplicate rows. `DataFrame.drop_duplicates()` returns the DataFrame that removes duplicate rows.

See [pandas.DataFrame.drop_duplicates](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html) for the important parameters.

>__Task 10__
>
>Remove duplicate rows and keep the first duplicate row

In [ ]:
...

Note that `drop_duplicates()` by default returns a non-duplicate DataFrame. However, the original DataFrame is not changed. If you would like to change the original DataFrame, you can add argument `inplace=True`.

We can also remove the rows where a specific column is duplicate by using `drop_duplicates(['columnName'])`.

>__Task 11__
>
>- Remove the rows if duplicates exist in columns: __age__, __job__, __marital__, __education__, __balance__
>- Keep the last duplicate rows
>- After you succeed with the above two steps, make the changes directly to the DataFrame

In [ ]:
...

In [ ]:
...